### PR TITLE
ngx.process.master_pid

### PR DESCRIPTION
**Pull request by chronolaw**
Here is the new API `ngx.process.master_pid`, which needs ffi C function `ngx_http_lua_ffi_master_pid` in ngx_lua.

I also write a test case in` t/process-master-pid.t`, But I am not familar with Perl and have not found the detail API documents of test::nginx, so I just copied `worker.t` then modified on it.

That test case may has error, please help me to correct it. Thanks very much.

### CODE REVIEW

**Comment by membphis** [comment]
you need to change the travis config, use your own branch to test.

https://github.com/openresty/lua-resty-core/blob/master/.travis.yml#L62

File: t/process-master-pid.t [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            for i = 1, 400 do
+                v = pid()
+            end
+            ngx.say(v)
+        }
+    }
+--- request
+GET /t
+--- response_body_like chop
+\d+$
```
Comment (agentzh) [inline t/process-master-pid.t]: This test is too loose. We could do better here by comparing the number returned by `master_pid()` against the real master pid saved in the nginx pid file.

File: t/process-master-pid.t [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+=== TEST 1: ngx.process.master_pid
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local process = require "ngx.process"
+
+            local v
+            local pid = process.master_pid
```
Comment (agentzh) [inline t/process-master-pid.t]: This is definitely wrong. You missed a pair of parentheses after `process.master_pid`. Have you actually run this test yourself?

This is another reason why you should make your PR pass the travis ci checks as @membphis suggested ;)

Comment (chronolaw) [reply t/process-master-pid.t]: Sorry, I'm not familar with travis ci, I will try to correct it ,thanks.

Comment (chronolaw) [reply t/process-master-pid.t]: And here is not an error ,because `pid` will be called in then next line ：
`local v = pid()`

**Comment by chronolaw** [comment]
I have run test in my own nginx 1.13.8+ngx_lua，and I will try to config travis ci after soon.

**Comment by chronolaw** [comment]
I have tried travis-ci build, but I got an error.

Then reason is that the `process.master_pid` needs nginx 1.13.8, but now OpenResty build will fail on it(no proper patch for it).

And in nginx 1.13.6, this API will return nil for no `ngx_parent` in nginx core.

Would you give me some advise to write the test case?

**Comment by chronolaw** [comment]
I created a new git branch to fit nginx 1.13.6, now `process.master_pid` passed ,but got another error
```
#   Failed test 'TEST 1: ngx.process.master_pid - pattern "(?^:\[TRACE   \d+ content_by_lua\(nginx\.conf:\d+\):4 loop\])" should match a line in error.log (req 0)'
#   at /home/travis/build/chronolaw/lua-resty-core/test-nginx/lib/Test/Nginx/Socket.pm line 1206.
#   Failed test 'TEST 1: ngx.process.master_pid - pattern "(?^:\[TRACE   \d+ content_by_lua\(nginx\.conf:\d+\):4 loop\])" should match a line in error.log (req 1)'
```
So could you tell my how to write the right `error_log eval` in xxx.t?

**Comment by chronolaw** [comment]
After some modify, my own test has passed, please see:https://travis-ci.org/chronolaw/lua-resty-core/jobs/324018825

Changes:
use `chronolaw/lua-nginx-module.git` and branch `test_ffi_master_pid`. In this branch code is 
```
#if (nginx_version >= 1013008)
    return (int) ngx_parent;
#else
    return (int) ((ngx_process == NGX_PROCESS_SINGLE) ? ngx_pid : getppid());
    //return -1;
#endif

```

**Comment by agentzh** [comment]
@chronolaw You can skip the tests for nginx cores older than 1.13.8 with the `--- skip_nginx` section. You can see such examples in this library's existing test suite.

**Comment by agentzh** [comment]
@chronolaw And you should point to your own fork of lua-nginx-module in `.travis.yml` in your branch, see others' PRs for such examples, like this:

https://github.com/openresty/lua-resty-core/pull/89

**Comment by agentzh** [comment]
Merged by fixing the aforementioned issues myself. Thanks!

**Pull request closed without merging**